python/vast/utils/arrow.py

```diff
@@ -1,7 +1,125 @@
+        num = num - (65535 << 32)
+    return ip.ip_address(num)
+
+
+class AddressScalar(pa.ExtensionScalar):
+    def as_py(self) -> ip.IPv4Address | ip.IPv6Address:
+        return unpack_ip(self.value.as_py())
+
+
+class AddressType(pa.ExtensionType):
+    ext_name = "vast.address_type"
+    ext_type = pa.binary(16)
+
+    def __init__(self):
+        pa.ExtensionType.__init__(self, self.ext_type, self.ext_name)
+
+    def __arrow_ext_serialize__(self) -> bytes:
+        return self.ext_name.encode()
+
+    @classmethod
+    def __arrow_ext_deserialize__(self, storage_type, serialized: bytes):
+        if serialized.decode() != self.ext_name:
+            raise TypeError("type identifier does not match")
+        if storage_type != self.ext_type:
+            raise TypeError("storage type does not match")
+
+    def __reduce__(self):
+        return AddressScalar, ()
+
+    def __arrow_ext_scalar_class__(self):
+        return AddressScalar
+
+
+class SubnetScalar(pa.ExtensionScalar):
+    def as_py(self) -> ip.IPv4Network | ip.IPv6Network:
+        address = unpack_ip(self.value[0].as_py())
+        length = self.value[1].as_py()
+        return ip.ip_network((address, length), strict=False)
+
+
+class SubnetType(pa.ExtensionType):
+    ext_name = "vast.subnet_type"
+    # This gives us a pyarrow.lib.ArrowNotImplementedError :-/
+    # ext_type = pa.struct([("address", AddressType()),
+    #                       ("length", pa.uint8())])
+    ext_type = pa.struct([("address", pa.binary(16)), ("length", pa.uint8())])
+
+    def __init__(self):
+        pa.ExtensionType.__init__(self, self.ext_type, self.ext_name)
+
+    def __arrow_ext_serialize__(self) -> bytes:
+        return self.ext_name.encode()
+
+    @classmethod
+    def __arrow_ext_deserialize__(self, storage_type, serialized: bytes):
+        if serialized.decode() != self.ext_name:
+            raise TypeError("type identifier does not match")
+        if storage_type != self.ext_type:
+            raise TypeError("storage type does not match")
+        return SubnetType()
+
+    def __reduce__(self):
+        return SubnetScalar, ()
+
+    def __arrow_ext_scalar_class__(self):
+        return SubnetScalar
+
+
+class EnumScalar(pa.ExtensionScalar):
+    def as_py(self) -> str:
+        return self.type.field(self.value.as_py())
+
+
+class EnumType(pa.ExtensionType):
+    ext_name = "vast.enumeration_type"
+    ext_type = pa.uint32()
+
+    def __init__(self, fields: dict[int, str]):
+        self._fields = fields
+        pa.ExtensionType.__init__(self, self.ext_type, self.ext_name)
+
+    @property
+    def fields(self):
+        return self._fields
+
+    def field(self, key):
+        return self._fields[key]
+
+    def __arrow_ext_serialize__(self) -> bytes:
+        return json.dumps(self._fields).encode()
+
+    @classmethod
+    def __arrow_ext_deserialize__(self, storage_type, serialized: bytes):
+        fields = json.loads(serialized.decode())
```

Comment:
I think you need to match VAST's serialization format exactly here:

Serialization:

https://github.com/tenzir/vast/blob/1c3eec37e96db254ad73a405164718b4d435d779/libvast/src/type.cpp#L2252-L2265

Deserialization:

https://github.com/tenzir/vast/blob/1c3eec37e96db254ad73a405164718b4d435d779/libvast/src/type.cpp#L2228-L2250

Reply:
I could only find this in the unit tests:

```cpp
  CHECK_EQUAL(fmt::format("{}", et), "enum {first: 0, third: 2, fourth: 3}");
```

It doesn't look like the output, given that it's invalid JSON. Do we have a valid JSON somewhere so that can make sure to get the *exact* representation?

Reply:
In 9cabcaa I reversed the field order during serialization, so that it's compatible with VAST.

I test that this generates valid JSON:

```python
    fields = {
        "foo": 1,
        "bar": 2,
        "baz": 3,
    }
    ty = vua.EnumType(fields)
    assert ty.__arrow_ext_serialize__().decode() == json.dumps(fields)
```

But we probably need a more thorough check by actually reading VAST data in the future.

Reply:
We have an `all_types.schema` file somewhere in the repository. If you create sample data of that, export it as Arrow, and then store that in some .bin file in the repository I think you have a good unit test setup.

Reply:
> I could only find this in the unit tests:
> 
> ```cpp
> CHECK_EQUAL(fmt::format("{}", et), "enum {first: 0, third: 2, fourth: 3}");
> ```
> 
> It doesn't look like the output, given that it's invalid JSON. Do we have a valid JSON somewhere so that can make sure to get the exact representation?

This is the printer of the enumeration type, it's not testing the Arrow extension type's serialization format. It's not related to it at all.

Reply:
we only test the ser-de cycle (write schema, read it back) and AFAICT do not have tests that actually look at the json.

There's an integration test looking at the output of `print_arrow.py` containing `ARROW:extension:metadata: '{ "A": 0, "B": 1, "C": 2 }'` over at `vast/integration/reference/arrow-full-data-model/step_01.ref`

Reply:
arion contains an IPC file dump that contains the `all-types.schema` mentioned above in `web/arion/tests/data/all-types_arrow-v2.ipc`. If we can read this from python, we should be good.

Reply:
Good call. I created a follow-up story to do this.

Resolving this thread because complete IPC support is currently still blocked anyway.